File: cli/terminal.ts
import chalk from 'chalk';
import inquirer from 'inquirer';
import ora from 'ora';
import { settings } from '../config/settings';
import type { Plan, Task } from '../types';

export class TerminalUI {
  private currentPlan: Plan | null = null;

  displayWelcome(): void {
    console.clear();
    console.log(chalk.cyan.bold('╔════════════════════════════════════════════╗'));
    console.log(chalk.cyan.bold('║'), chalk.yellow.bold(' GLM Code System ').padEnd(54), chalk.cyan.bold('║'));
    console.log(chalk.cyan.bold('║'), chalk.white('Autonomous coding with self-learning capabilities').padEnd(54), chalk.cyan.bold('║'));
    console.log(chalk.cyan.bold('╚════════════════════════════════════════════╝'));
    console.log();
  }

  displayAgentThought(agentName: string, thought: string): void {
    console.log(chalk.blue.bold(`[${agentName}]`), chalk.blue(thought));
  }

  displayPlan(plan: Plan): void {
    this.currentPlan = plan;
    console.log();
    console.log(chalk.green.bold('━━━━━━━━━━━━━━'));
    console.log(chalk.green.bold('Development Plan'));
    console.log(chalk.green.bold('━━━━━━━━━━━━━━')));
    console.log(plan.plan);

    if (plan.relevantPatterns && plan.relevantPatterns.length > 0) {
      console.log();
      console.log(chalk.bold('Relevant Patterns:'));
      for (const pattern of plan.relevantPatterns) {
        const successRate = (pattern.successRate * 100).toFixed(0);
        console.log(
          chalk.cyan(`  - ${pattern.type}:`),
          chalk.white(pattern.description),
          chalk.yellow(` (${successRate}% success rate)`)
        );
      }
    }
    console.log();
  }

  displayTaskStart(task: Task, index: number, total: number): void {
    console.log();
    console.log(
      chalk.yellow.bold(`Task ${index}/${total}:`),
      chalk.white(task.description),
      chalk.cyan(` (${task.complexity})`)
    );
  }

  displayTaskComplete(task: Task, success: boolean, output: string): void {
    const status = success ? chalk.green.bold('✓') : chalk.red.bold('✗');
    console.log(`${status} ${task.description}`);

    if (success && output) {
      const preview = output.substring(0, 200);
      console.log(chalk.dim(preview + (output.length > 200 ? '...' : '')));
    }
  }

  displayProgress(current: number, total: number, message: string = 'Processing...'): void {
    const spinner = ora(message).start();
    const percentage = Math.round((current / total) * 100);
    spinner.text = `${message} ${percentage}%`;
    spinner.succeed();
  }

  async displayStreamingThought(agentName: string, stream: AsyncGenerator<string>): Promise<void> {
    process.stdout.write(chalk.blue.bold(`[${agentName}] `) + ' ');

    for await (const chunk of stream) {
      process.stdout.write(chunk);
    }

    process.stdout.write('\n');
  }

  displayError(error: string): void {
    console.log();
    console.log(chalk.red.bold('Error:'), chalk.red(error));
  }

  displaySuccess(message: string): void {
    console.log();
    console.log(chalk.green.bold('Success:'), chalk.green(message));
  }

  displayLearning(patternsCount: number): void {
    if (patternsCount > 0) {
      console.log();
      console.log(
        chalk.cyan.bold('[Learning]'),
        chalk.white(`Recorded ${patternsCount} new patterns in knowledge base`)
      );
    }
  }

  displayMetrics(metrics: any): void {
    const successRate =
      metrics.totalTasks > 0
        ? (metrics.successfulTasks / metrics.totalTasks)
        : 0;

    console.log();
    console.log(chalk.cyan.bold('System Metrics'));
    console.log(chalk.cyan('━━━━━━━━━━━━━━'));
    console.log(chalk.white(`Total Tasks: ${metrics.totalTasks}`));
    console.log(chalk.white(`Successful: ${metrics.successfulTasks}`));
    console.log(chalk.yellow(`Success Rate: ${(successRate * 100).toFixed(1)}%`));
    console.log(chalk.white(`Patterns Learned: ${metrics.patternsLearned}`));
    console.log();
  }

  displayKnowledge(patterns: any[]): void {
    if (patterns.length === 0) {
      console.log(chalk.yellow('No patterns learned yet.'));
      return;
    }

    console.log();
    console.log(chalk.bold('Learned Patterns:'));
    for (let i = 0; i < Math.min(patterns.length, 10); i++) {
      const pattern = patterns[i];
      const successRate = (pattern.successRate * 100).toFixed(0);
      console.log();
      console.log(chalk.cyan(`${i + 1}.`), chalk.bold(pattern.type));
      console.log(chalk.white(`  (${successRate}%) used ${pattern.usageCount || 0} times`));
      console.log(chalk.dim(`   ${pattern.description}`));
    }

    if (patterns.length > 10) {
      console.log();
      console.log(chalk.dim(`... and ${patterns.length - 10} more patterns`));
    }
  }

  async displayUserInput(): Promise<string> {
    const answer = await inquirer.prompt([
      {
        type: 'input',
        name: 'command',
        message: chalk.cyan('glm-code>'),
      },
    ]);

    return answer.command;
  }

  async displayConfirmPlan(): Promise<boolean> {
    const answer = await inquirer.prompt([
      {
        type: 'confirm',
        name: 'confirm',
        message: 'Execute this plan?',
        default: true,
      },
    ]);

    return answer.confirm;
  }

  displayHelp(): void {
    console.log();
    console.log(chalk.bold('Available Commands:'));
    console.log();
    console.log(chalk.cyan('/help'), chalk.white('   - Show help information'));
    console.log(chalk.cyan('/plan'), chalk.white('   - Display current plan'));
    console.log(chalk.cyan('/status'), chalk.white('   - Show system status and metrics'));
    console.log(chalk.cyan('/learn'), chalk.white('   - Display learned patterns'));
    console.log(chalk.cyan('/feedback'), chalk.white('   - Provide feedback on system performance'));
    console.log(chalk.cyan('/clear'), chalk.white('   - Clear conversation history'));
    console.log(chalk.cyan('/quit'), chalk.white('   - Exit system'));
    console.log();
  }
}

export type AsyncGenerator<T> = {
  [Symbol.asyncIterator](): AsyncIterator<T>;
}

File: glm_code_system/cli/terminal.py
"""Terminal UI for GLM Code System."""

from rich.console import Console
from rich.panel import Panel
from rich.progress import Progress, SpinnerColumn, TextColumn, BarColumn
from rich.text import Text
from typing import Any

console = Console()


class TerminalUI:
    """Terminal user interface for the system."""

    def __init__(self) -> None:
        """Initialize terminal UI."""
        self.console = console

    def display_welcome(self) -> None:
        """Display welcome message."""
        welcome_text = """
[bold cyan]GLM Code System[/bold cyan]
[cyan]Autonomous coding with self-learning capabilities[cyan]

Type your task or /[command] to interact.
Available commands:
  /plan     - Show current plan
  /status   - Show system status
  /learn    - Show learned patterns
  /help     - Display help
  /quit     - Exit system
        """
        self.console.print(Panel(welcome_text, title="Welcome", border_style="cyan"))

    def display_agent_thought(self, agent_name: str, thought: str) -> None:
        """Display agent thought process."""
        self.console.print(
            f"\n[bold blue][{agent_name}][/bold blue] {thought}"
        )

    def display_plan(self, plan: dict[str, Any]) -> None:
        """Display development plan."""
        self.console.print(
            Panel(
                plan["plan"],
                title="[bold green]Development Plan[/bold green]",
                border_style="green",
            )
        )

        if plan.get("relevant_patterns"):
            self.console.print("\n[bold]Relevant Patterns:[/bold]")
            for pattern in plan["relevant_patterns"]:
                self.console.print(
                    f"  - [cyan]{pattern['type']}[/cyan]: {pattern['description']} "
                    f"([yellow]{pattern['success_rate']:.0%}[/yellow] success rate)"
                )

    def display_task_start(self, task: dict[str, Any], index: int, total: int) -> None:
        """Display start of task execution."""
        self.console.print(
            f"\n[bold yellow]Task {index}/{total}[/bold yellow]: "
            f"{task['description']} ([cyan]{task.get('complexity', 'medium')}[/cyan])"
        )

    def display_task_complete(
        self,
        task: dict[str, Any],
        success: bool,
        output: str,
    ) -> None:
        """Display task completion."""
        status = "[bold green]✓[/bold green]" if success else "[bold red]✗[/bold red]"
        self.console.print(
            f"{status} {task['description']}",
        )

        if success and output:
            self.console.print(f"\n[dim]{output[:200]}...[/dim]")

    def display_progress(
        self,
        current: int,
        total: int,
        message: str = "Processing...",
    ) -> None:
        """Display progress bar."""
        with Progress(
            SpinnerColumn(),
            TextColumn("[progress.description]{task.description}"),
            BarColumn(),
            TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
            console=self.console,
        ) as progress:
            progress.add_task(message, total=total, completed=current)

    async def display_streaming_thought(self, agent_name: str, stream) -> None:
        """Display streaming agent thought."""
        self.console.print(f"\n[bold blue][{agent_name}][/bold blue] ", end="")

        async for chunk in stream:
            self.console.print(chunk, end="")

        self.console.print()

    def display_error(self, error: str) -> None:
        """Display error message."""
        self.console.print(
            f"\n[bold red]Error:[/bold red] {error}",
        )

    def display_success(self, message: str) -> None:
        """Display success message."""
        self.console.print(
            f"\n[bold green]Success:[/bold green] {message}",
        )

    def display_learning(self, patterns_count: int) -> None:
        """Display learning progress."""
        if patterns_count > 0:
            self.console.print(
                f"\n[bold cyan][Learning][/bold cyan] "
                f"Recorded {patterns_count} new patterns in knowledge base"
            )

    def display_metrics(
        self,
        metrics: dict[str, Any],
    ) -> None:
        """Display performance metrics."""
        success_rate = 0
        total = metrics.get("total_tasks", 0)
        if total > 0:
            success_rate = metrics["successful_tasks"] / total

        metrics_text = f"""
[bold]System Metrics[/bold]
━━━━━━━━━━━━━━━━
Total Tasks: {total}
Successful: {metrics.get('successful_tasks', 0)}
Success Rate: [cyan]{success_rate:.1%}[/cyan]
Patterns Learned: {metrics.get('patterns_learned', 0)}
        """
        self.console.print(Panel(metrics_text, border_style="cyan"))

    def display_user_input(self) -> str:
        """Get user input with prompt."""
        return input("\n[bold cyan]glm-code>[/bold cyan] ")

    def display_help(self) -> None:
        """Display help information."""
        help_text = """
[bold]Available Commands:[/bold]
  /[command] - Execute command

[bold]Commands:[/bold]
  /plan     - Show current development plan
  /status   - Show system status and metrics
  /learn    - Display learned patterns and knowledge
  /feedback - Provide feedback on system performance
  /clear    - Clear conversation history
  /help     - Show this help message
  /quit     - Exit the system

[bold]Examples:[/bold]
  Create a user authentication system
  Fix the bug in api/users.py
  Add tests for the payment module
        """
        self.console.print(Panel(help_text, title="Help", border_style="cyan"))

    def display_knowledge(self, patterns: list[Any]) -> None:
        """Display learned patterns."""
        if not patterns:
            self.console.print("\n[yellow]No patterns learned yet.[/yellow]")
            return

        self.console.print("\n[bold]Learned Patterns:[/bold]")
        for i, pattern in enumerate(patterns[:10], 1):
            self.console.print(
                f"\n{[cyan]{i}.[/cyan] [bold]{pattern['type']}[/bold] "
                f"([yellow]{pattern['success_rate']:.0%}[/yellow] used {pattern.get('usage_count', 0)} times)"
            )
            self.console.print(f"   {pattern['description']}")

        if len(patterns) > 10:
            self.console.print(f"\n... and {len(patterns) - 10} more patterns")

    def clear_screen(self) -> None:
        """Clear terminal screen."""
        self.console.clear()
